### PR TITLE
Precise logging

### PR DESCRIPTION
**Pull request by kennymckormick**
Make logger print average loss, acc, etc. over multiple GPUS when using distributed training, which is more meaningful.

### CODE REVIEW

**Comment by hellock** [comment]
Unit tests fail.

File: mmcv/runner/hooks/logger/base.py [hunk 1]
```diff
@@ -37,23 +43,46 @@ def before_epoch(self, runner):
...
-            runner.log_buffer.clear_output()
+        self.sync_buffer_output(runner)
+        if runner.log_buffer.ready:
+            self.log(runner)
+            if self.reset_flag:
+                runner.log_buffer.clear_output()
+
+    def sync_buffer_output(self, runner):
+        for k, v in runner.log_buffer.output.items():
+            tmp_tensor = torch.Tensor([v]).cuda(torch.cuda.current_device())
```
Comment (xvjiarui) [inline mmcv/runner/hooks/logger/base.py]: Check if cuda is available.

File: mmcv/runner/hooks/logger/base.py [hunk 2]
```diff
@@ -20,6 +23,9 @@ def __init__(self, interval=10, ignore_last=True, reset_flag=False):
         self.interval = interval
         self.ignore_last = ignore_last
         self.reset_flag = reset_flag
+        _rank, _world_size = get_dist_info()
+        self.rank = _rank
+        self.world_size = _world_size
```
Comment (xvjiarui) [inline mmcv/runner/hooks/logger/base.py]: `rank` and `world_size` could be accessed from `runner` when needed.

File: mmcv/runner/hooks/logger/base.py [hunk 3]
```diff
@@ -37,23 +43,46 @@ def before_epoch(self, runner):
...
         elif self.end_of_epoch(runner) and not self.ignore_last:
-            # not precise but more stable
             runner.log_buffer.average(self.interval)
+            self.sync_buffer_output(runner)
+        if runner.log_buffer.ready:
+            self.log(runner)
+            if self.reset_flag:
+                runner.log_buffer.clear_output()
 
+    def after_val_iter(self, runner):
```
Comment (xvjiarui) [inline mmcv/runner/hooks/logger/base.py]: Is it necessary to log during validation? If so, please just implement `after_iter`.

File: mmcv/runner/hooks/logger/base.py [hunk 4]
```diff
@@ -37,23 +43,46 @@ def before_epoch(self, runner):
...
+        self.sync_buffer_output(runner)
+        if runner.log_buffer.ready:
+            self.log(runner)
+            if self.reset_flag:
+                runner.log_buffer.clear_output()
+
+    def sync_buffer_output(self, runner):
+        for k, v in runner.log_buffer.output.items():
+            tmp_tensor = torch.Tensor([v]).cuda(torch.cuda.current_device())
+            dist.all_reduce(tmp_tensor)
```
Comment (xvjiarui) [inline mmcv/runner/hooks/logger/base.py]: Need to check if cuda or dist is available

File: mmcv/runner/hooks/logger/base.py [hunk 5]
```diff
@@ -37,23 +43,46 @@ def before_epoch(self, runner):
...
             if self.reset_flag:
                 runner.log_buffer.clear_output()
 
     def after_val_epoch(self, runner):
         runner.log_buffer.average()
-        self.log(runner)
-        if self.reset_flag:
-            runner.log_buffer.clear_output()
+        self.sync_buffer_output(runner)
+        if runner.log_buffer.ready:
```
Comment (xvjiarui) [inline mmcv/runner/hooks/logger/base.py]: Maybe there is no need to check `runner.log_buffer.ready` after `runner.log_buffer.average()`?

**Pull request closed without merging**